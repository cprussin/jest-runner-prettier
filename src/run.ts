import { TestResult } from "@jest/test-result";
import { emphasize } from "emphasize";
import { pass, fail } from "create-jest-runner";
import fs from "fs";
import { diff } from "jest-diff";
import prettier from "prettier";

interface Parameters {
  testPath: string;
}

export default ({ testPath }: Parameters): Promise<TestResult> => {
  const start = new Date().getTime();
  const contents = fs.readFileSync(testPath, "utf8");

  return prettier.resolveConfig(testPath).then((config) => {
    const prettierConfig = Object.assign({}, config, {
      filepath: testPath,
    });

    const isPretty = prettier.check(contents, prettierConfig);
    if (isPretty) {
      return pass({
        start,
        end: new Date().getTime(),
        test: { path: testPath },
      });
    }

    const formatted = prettier.format(contents, prettierConfig);

    return fail({
      start,
      end: new Date().getTime(),
      test: {
        path: testPath,
        errorMessage: diff(
          emphasize.highlightAuto(formatted).value,
          emphasize.highlightAuto(contents).value,
          {
            expand: false,
          }
        ),
      },
    });
  });
};
