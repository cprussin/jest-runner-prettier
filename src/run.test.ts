import path from "node:path";
import run from "./run";
import { fileURLToPath } from "node:url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Remove undeterministic data from test reports
expect.addSnapshotSerializer({
  print: (value: any, serialize) => {
    delete value.perfStats;
    delete value.testFilePath;
    for (const result of value.testResults) {
      delete result.duration;
    }
    return serialize(value);
  },
  test: (value) =>
    value && value.perfStats && value.testFilePath && value.testResults,
});

describe("jest-runner-prettier", () => {
  describe("JSON", () => {
    describe("good fixture", () => {
      it("matches snapshot", () => {
        return run({
          testPath: path.join(__dirname, "__fixtures__", `good.json`),
        }).then((result) => expect(result).toMatchSnapshot());
      });
    });

    describe("bad fixture", () => {
      it("matches snapshot", () => {
        return run({
          testPath: path.join(__dirname, "__fixtures__", `bad.json`),
        }).then((result) => expect(result).toMatchSnapshot());
      });
    });
  });

  describe("JSX", () => {
    describe("good fixture", () => {
      it("matches snapshot", () => {
        return run({
          testPath: path.join(__dirname, "__fixtures__", `good.jsx`),
        }).then((result) => expect(result).toMatchSnapshot());
      });
    });

    describe("bad fixture", () => {
      it("matches snapshot", () => {
        return run({
          testPath: path.join(__dirname, "__fixtures__", `bad.jsx`),
        }).then((result) => expect(result).toMatchSnapshot());
      });
    });
  });
});
