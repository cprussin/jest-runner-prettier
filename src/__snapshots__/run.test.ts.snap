// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jest-runner-prettier JSON bad fixture matches snapshot 1`] = `
Object {
  "console": null,
  "failureMessage": "[32m- Expected[39m
[31m+ Received[39m

[32m- {[39m
[32m-   [33m\\"ohMyGodLongArray\\"[39m[32m: [[39m
[32m-     [36m\\"prettyLongValue\\"[39m[32m,[39m
[32m-     [36m\\"decentlyLongValue\\"[39m[32m,[39m
[32m-     [36m\\"prettyDecentlyLongValue\\"[39m[32m,[39m
[32m-     [36m\\"wowWowWowWhatALongValue\\"[39m[32m[39m
[32m-   ],[39m
[32m-   [33m\\"woahThisObjectIsDisapperingOffTheScreen\\"[39m[32m: {[39m
[32m-     [33m\\"foo\\"[39m[32m: [36m\\"bar\\"[39m[32m,[39m
[32m-     [33m\\"bar\\"[39m[32m: [36m\\"foo\\"[39m[32m,[39m
[32m-     [33m\\"faz\\"[39m[32m: [32mfalse[39m[32m[39m
[32m-   },[39m
[32m-   [33m\\"truthy\\"[39m[32m: [32mtrue[39m[32m,[39m
[32m-   [33m\\"aNumber\\"[39m[32m: [36m123[39m[32m,[39m
[32m-   [33m\\"aString\\"[39m[32m: [36m\\"hi\\"[39m[32m[39m
[32m- }[39m
[32m-[39m
[31m+ {[33m\\"ohMyGodLongArray\\"[39m[31m: [[36m\\"prettyLongValue\\"[39m[31m, [36m\\"decentlyLongValue\\"[39m[31m,[39m
[31m+   [36m\\"prettyDecentlyLongValue\\"[39m[31m, [36m\\"wowWowWowWhatALongValue\\"[39m[31m], [33m\\"woahThisObjectIsDisapperingOffTheScreen\\"[39m[31m: {[33m\\"foo\\"[39m[31m: [36m\\"bar\\"[39m[31m,[39m
[31m+     [33m\\"bar\\"[39m[31m: [36m\\"foo\\"[39m[31m, [33m\\"faz\\"[39m[31m: [32mfalse[39m[31m}, [33m\\"truthy\\"[39m[31m: [32mtrue[39m[31m, [33m\\"aNumber\\"[39m[31m: [36m123[39m[31m, [33m\\"aString\\"[39m[31m: [36m\\"hi\\"[39m[31m[39m
[31m+   }[39m",
  "numFailingTests": 1,
  "numPassingTests": 0,
  "numPendingTests": 0,
  "numTodoTests": 0,
  "skipped": undefined,
  "snapshot": Object {
    "added": 0,
    "fileDeleted": false,
    "matched": 0,
    "unchecked": 0,
    "unmatched": 0,
    "updated": 0,
  },
  "sourceMaps": Object {},
  "testExecError": null,
  "testResults": Array [
    Object {
      "ancestorTitles": Array [],
      "failureMessages": Array [
        "[32m- Expected[39m
[31m+ Received[39m

[32m- {[39m
[32m-   [33m\\"ohMyGodLongArray\\"[39m[32m: [[39m
[32m-     [36m\\"prettyLongValue\\"[39m[32m,[39m
[32m-     [36m\\"decentlyLongValue\\"[39m[32m,[39m
[32m-     [36m\\"prettyDecentlyLongValue\\"[39m[32m,[39m
[32m-     [36m\\"wowWowWowWhatALongValue\\"[39m[32m[39m
[32m-   ],[39m
[32m-   [33m\\"woahThisObjectIsDisapperingOffTheScreen\\"[39m[32m: {[39m
[32m-     [33m\\"foo\\"[39m[32m: [36m\\"bar\\"[39m[32m,[39m
[32m-     [33m\\"bar\\"[39m[32m: [36m\\"foo\\"[39m[32m,[39m
[32m-     [33m\\"faz\\"[39m[32m: [32mfalse[39m[32m[39m
[32m-   },[39m
[32m-   [33m\\"truthy\\"[39m[32m: [32mtrue[39m[32m,[39m
[32m-   [33m\\"aNumber\\"[39m[32m: [36m123[39m[32m,[39m
[32m-   [33m\\"aString\\"[39m[32m: [36m\\"hi\\"[39m[32m[39m
[32m- }[39m
[32m-[39m
[31m+ {[33m\\"ohMyGodLongArray\\"[39m[31m: [[36m\\"prettyLongValue\\"[39m[31m, [36m\\"decentlyLongValue\\"[39m[31m,[39m
[31m+   [36m\\"prettyDecentlyLongValue\\"[39m[31m, [36m\\"wowWowWowWhatALongValue\\"[39m[31m], [33m\\"woahThisObjectIsDisapperingOffTheScreen\\"[39m[31m: {[33m\\"foo\\"[39m[31m: [36m\\"bar\\"[39m[31m,[39m
[31m+     [33m\\"bar\\"[39m[31m: [36m\\"foo\\"[39m[31m, [33m\\"faz\\"[39m[31m: [32mfalse[39m[31m}, [33m\\"truthy\\"[39m[31m: [32mtrue[39m[31m, [33m\\"aNumber\\"[39m[31m: [36m123[39m[31m, [33m\\"aString\\"[39m[31m: [36m\\"hi\\"[39m[31m[39m
[31m+   }[39m",
      ],
      "fullName": undefined,
      "numPassingAsserts": 1,
      "status": "failed",
      "title": "",
    },
  ],
}
`;

exports[`jest-runner-prettier JSON good fixture matches snapshot 1`] = `
Object {
  "console": null,
  "failureMessage": undefined,
  "numFailingTests": 0,
  "numPassingTests": 1,
  "numPendingTests": 0,
  "numTodoTests": 0,
  "skipped": undefined,
  "snapshot": Object {
    "added": 0,
    "fileDeleted": false,
    "matched": 0,
    "unchecked": 0,
    "unmatched": 0,
    "updated": 0,
  },
  "sourceMaps": Object {},
  "testExecError": null,
  "testResults": Array [
    Object {
      "ancestorTitles": Array [],
      "failureMessages": Array [],
      "fullName": undefined,
      "numPassingAsserts": 0,
      "status": "passed",
      "title": "",
    },
  ],
}
`;

exports[`jest-runner-prettier JSX bad fixture matches snapshot 1`] = `
Object {
  "console": null,
  "failureMessage": "[32m- Expected[39m
[31m+ Received[39m

[32m- [32mfunction[39m[32m [34mHelloWorld[39m[32m({[39m
[32m-   greeting = [36m\\"hello\\"[39m[32m,[39m
[32m-   greeted = [36m'\\"World\\"'[39m[32m,[39m
[32m-   silent = [36mfalse[39m[32m,[39m
[32m-   onMouseOver,[39m
[32m- }) {[39m
[32m-   [32mif[39m[32m (!greeting) {[39m
[32m-     [32mreturn[39m[32m [36mnull[39m[32m;[39m
[32m-   }[39m
[31m+ [32mfunction[39m[31m [34mHelloWorld[39m[31m({greeting = [36m\\"hello\\"[39m[31m, greeted = [36m'\\"World\\"'[39m[31m, silent = [36mfalse[39m[31m, onMouseOver,}) {[39m
[31m+[39m
[31m+   [32mif[39m[31m(!greeting){[32mreturn[39m[31m [36mnull[39m[31m};[39m
[31m+[39m
[31m+      [90m// [36mTODO:[39m[31m[90m Don't use random in render[39m[31m[39m
[31m+   [32mlet[39m[31m num = [34mMath[39m[31m.floor ([34mMath[39m[31m.[34mrandom[39m[31m() * [36m1E+7[39m[31m).[34mtoString[39m[31m().[34mreplace[39m[31m([36m/\\\\.\\\\d+/ig[39m[31m, [36m\\"\\"[39m[31m)[39m
[31m+[39m
[31m+   [32mreturn[39m[31m <[34mdiv[39m[31m [33mclassName[39m[31m=[36m'HelloWorld'[39m[31m [33mtitle[39m[31m=[36m{[39m[31m\`[33mYou[39m[31m [33mare[39m[31m [33mvisitor[39m[31m [33mnumber[39m[31m \${ [33mnum[39m[31m }\`} [33monMouseOver[39m[31m=[36m{onMouseOver}[39m[31m>[39m
[31m+[39m
[31m+     <[34mstrong[39m[31m>{ greeting.slice( 0, 1 ).toUpperCase() + greeting.slice(1).toLowerCase() }</[34mstrong[39m[31m>[39m
[31m+     {greeting.endsWith(\\",\\") ? \\" \\" : <[34mspan[39m[31m [33mstyle[39m[31m=[36m{{color:[39m[31m '\\\\[33mgrey[39m[31m'}}>\\", \\"</[34mspan[39m[31m> }[39m
[31m+     <[34mem[39m[31m>[39m
[31m+ 	{ greeted }[39m
[31m+ 	</[34mem[39m[31m>[39m
[31m+     { (silent)[39m
[31m+       ? \\".\\"[39m
[31m+       : \\"!\\"}[39m

[32m-   [90m// [36mTODO:[39m[32m[90m Don't use random in render[39m[32m[39m
[32m-   [32mlet[39m[32m num = [34mMath[39m[32m.[34mfloor[39m[32m([34mMath[39m[32m.[34mrandom[39m[32m() * [36m1e7[39m[32m)[39m
[32m-     .[34mtoString[39m[32m()[39m
[32m-     .[34mreplace[39m[32m([36m/\\\\.\\\\d+/gi[39m[32m, [36m\\"\\"[39m[32m);[39m
[31m+     </[34mdiv[39m[31m>;[39m

[32m-   [32mreturn[39m[32m ([39m
[32m-     <[34mdiv[39m[32m[39m
[32m-       [33mclassName[39m[32m=[36m\\"HelloWorld\\"[39m[32m[39m
[32m-       [33mtitle[39m[32m=[36m{[39m[32m\`[33mYou[39m[32m [33mare[39m[32m [33mvisitor[39m[32m [33mnumber[39m[32m \${[33mnum[39m[32m}\`}[39m
[32m-       [33monMouseOver[39m[32m=[36m{onMouseOver}[39m[32m[39m
[32m-     >[39m
[32m-       <[34mstrong[39m[32m>[39m
[32m-         {greeting.slice(0, 1).toUpperCase() + greeting.slice(1).toLowerCase()}[39m
[32m-       </[34mstrong[39m[32m>[39m
[32m-       {greeting.endsWith(\\",\\") ? ([39m
[32m-         \\" \\"[39m
[32m-       ) : ([39m
[32m-         <[34mspan[39m[32m [33mstyle[39m[32m=[36m{{[39m[32m [33mcolor:[39m[32m \\"[33mgrey[39m[32m\\" }}>\\", \\"</[34mspan[39m[32m>[39m
[32m-       )}[39m
[32m-       <[34mem[39m[32m>{greeted}</[34mem[39m[32m>[39m
[32m-       {silent ? \\".\\" : \\"!\\"}[39m
[32m-     </[34mdiv[39m[32m>[39m
[32m-   );[39m
[2m  }[22m
[32m-[39m",
  "numFailingTests": 1,
  "numPassingTests": 0,
  "numPendingTests": 0,
  "numTodoTests": 0,
  "skipped": undefined,
  "snapshot": Object {
    "added": 0,
    "fileDeleted": false,
    "matched": 0,
    "unchecked": 0,
    "unmatched": 0,
    "updated": 0,
  },
  "sourceMaps": Object {},
  "testExecError": null,
  "testResults": Array [
    Object {
      "ancestorTitles": Array [],
      "failureMessages": Array [
        "[32m- Expected[39m
[31m+ Received[39m

[32m- [32mfunction[39m[32m [34mHelloWorld[39m[32m({[39m
[32m-   greeting = [36m\\"hello\\"[39m[32m,[39m
[32m-   greeted = [36m'\\"World\\"'[39m[32m,[39m
[32m-   silent = [36mfalse[39m[32m,[39m
[32m-   onMouseOver,[39m
[32m- }) {[39m
[32m-   [32mif[39m[32m (!greeting) {[39m
[32m-     [32mreturn[39m[32m [36mnull[39m[32m;[39m
[32m-   }[39m
[31m+ [32mfunction[39m[31m [34mHelloWorld[39m[31m({greeting = [36m\\"hello\\"[39m[31m, greeted = [36m'\\"World\\"'[39m[31m, silent = [36mfalse[39m[31m, onMouseOver,}) {[39m
[31m+[39m
[31m+   [32mif[39m[31m(!greeting){[32mreturn[39m[31m [36mnull[39m[31m};[39m
[31m+[39m
[31m+      [90m// [36mTODO:[39m[31m[90m Don't use random in render[39m[31m[39m
[31m+   [32mlet[39m[31m num = [34mMath[39m[31m.floor ([34mMath[39m[31m.[34mrandom[39m[31m() * [36m1E+7[39m[31m).[34mtoString[39m[31m().[34mreplace[39m[31m([36m/\\\\.\\\\d+/ig[39m[31m, [36m\\"\\"[39m[31m)[39m
[31m+[39m
[31m+   [32mreturn[39m[31m <[34mdiv[39m[31m [33mclassName[39m[31m=[36m'HelloWorld'[39m[31m [33mtitle[39m[31m=[36m{[39m[31m\`[33mYou[39m[31m [33mare[39m[31m [33mvisitor[39m[31m [33mnumber[39m[31m \${ [33mnum[39m[31m }\`} [33monMouseOver[39m[31m=[36m{onMouseOver}[39m[31m>[39m
[31m+[39m
[31m+     <[34mstrong[39m[31m>{ greeting.slice( 0, 1 ).toUpperCase() + greeting.slice(1).toLowerCase() }</[34mstrong[39m[31m>[39m
[31m+     {greeting.endsWith(\\",\\") ? \\" \\" : <[34mspan[39m[31m [33mstyle[39m[31m=[36m{{color:[39m[31m '\\\\[33mgrey[39m[31m'}}>\\", \\"</[34mspan[39m[31m> }[39m
[31m+     <[34mem[39m[31m>[39m
[31m+ 	{ greeted }[39m
[31m+ 	</[34mem[39m[31m>[39m
[31m+     { (silent)[39m
[31m+       ? \\".\\"[39m
[31m+       : \\"!\\"}[39m

[32m-   [90m// [36mTODO:[39m[32m[90m Don't use random in render[39m[32m[39m
[32m-   [32mlet[39m[32m num = [34mMath[39m[32m.[34mfloor[39m[32m([34mMath[39m[32m.[34mrandom[39m[32m() * [36m1e7[39m[32m)[39m
[32m-     .[34mtoString[39m[32m()[39m
[32m-     .[34mreplace[39m[32m([36m/\\\\.\\\\d+/gi[39m[32m, [36m\\"\\"[39m[32m);[39m
[31m+     </[34mdiv[39m[31m>;[39m

[32m-   [32mreturn[39m[32m ([39m
[32m-     <[34mdiv[39m[32m[39m
[32m-       [33mclassName[39m[32m=[36m\\"HelloWorld\\"[39m[32m[39m
[32m-       [33mtitle[39m[32m=[36m{[39m[32m\`[33mYou[39m[32m [33mare[39m[32m [33mvisitor[39m[32m [33mnumber[39m[32m \${[33mnum[39m[32m}\`}[39m
[32m-       [33monMouseOver[39m[32m=[36m{onMouseOver}[39m[32m[39m
[32m-     >[39m
[32m-       <[34mstrong[39m[32m>[39m
[32m-         {greeting.slice(0, 1).toUpperCase() + greeting.slice(1).toLowerCase()}[39m
[32m-       </[34mstrong[39m[32m>[39m
[32m-       {greeting.endsWith(\\",\\") ? ([39m
[32m-         \\" \\"[39m
[32m-       ) : ([39m
[32m-         <[34mspan[39m[32m [33mstyle[39m[32m=[36m{{[39m[32m [33mcolor:[39m[32m \\"[33mgrey[39m[32m\\" }}>\\", \\"</[34mspan[39m[32m>[39m
[32m-       )}[39m
[32m-       <[34mem[39m[32m>{greeted}</[34mem[39m[32m>[39m
[32m-       {silent ? \\".\\" : \\"!\\"}[39m
[32m-     </[34mdiv[39m[32m>[39m
[32m-   );[39m
[2m  }[22m
[32m-[39m",
      ],
      "fullName": undefined,
      "numPassingAsserts": 1,
      "status": "failed",
      "title": "",
    },
  ],
}
`;

exports[`jest-runner-prettier JSX good fixture matches snapshot 1`] = `
Object {
  "console": null,
  "failureMessage": undefined,
  "numFailingTests": 0,
  "numPassingTests": 1,
  "numPendingTests": 0,
  "numTodoTests": 0,
  "skipped": undefined,
  "snapshot": Object {
    "added": 0,
    "fileDeleted": false,
    "matched": 0,
    "unchecked": 0,
    "unmatched": 0,
    "updated": 0,
  },
  "sourceMaps": Object {},
  "testExecError": null,
  "testResults": Array [
    Object {
      "ancestorTitles": Array [],
      "failureMessages": Array [],
      "fullName": undefined,
      "numPassingAsserts": 0,
      "status": "passed",
      "title": "",
    },
  ],
}
`;
